sound/soc/sof/sof-audio.c

```diff
@@ -107,6 +107,17 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 	if (!swidget->private)
 		return 0;
 
+	/*
+	 * PCM streams can be freed without being restarted after system resume. Prevent
+	 * the use_count from being decremented below 0 for dynamic widgets and below 1
+	 * for static widgets in such scenarios.
+	 */
+	if (swidget->dynamic_pipeline_widget && !swidget->use_count)
+			return 0;
+
+	if (!swidget->dynamic_pipeline_widget && swidget->use_count == 1)
+			return 0;
+
```

Comment:
maybe a side question, but sof_widget_free() is also used from intel/hda.c

````
int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w, bool setup)
{
	struct snd_sof_widget *swidget = w->dobj.private;
	struct snd_soc_component *component = swidget->scomp;
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
	struct sof_ipc_dai_config *config;
	struct snd_sof_dai *sof_dai;
	struct sof_ipc_reply reply;
	int ret;

	sof_dai = swidget->private;

	if (!sof_dai || !sof_dai->dai_config) {
		dev_err(sdev->dev, "No config for DAI %s\n", w->name);
		return -EINVAL;
	}

	config = &sof_dai->dai_config[sof_dai->current_config];

	/*
	 * For static pipelines, the DAI widget would already be set up and calling
	 * sof_widget_setup()/free() simply returns without doing anything.
	 * For dynamic pipelines, the DAI widget will be set up or freed here.
	 */
	if (setup) {
		ret = sof_widget_setup(sdev, swidget);
		if (ret < 0) {
			dev_err(sdev->dev, "error: setting up DAI widget %s\n", w->name);
			return ret;
		}

		/* send DAI_CONFIG IPC */
		return sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
					  &reply, sizeof(reply));
	}

	ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
				  &reply, sizeof(reply));
	if (ret < 0) {
		dev_err(sdev->dev, "error: updating DAI config for %s\n", w->name);
		return ret;
	}

	return sof_widget_free(sdev, swidget);
}
````

would the change impact the HDaudio DAI config?